### PR TITLE
KOGITO-3716 Allow automatic close of staging repo

### PR DESCRIPTION
**Pull request by radtriste**
https://issues.redhat.com/browse/KOGITO-3716

Dependent PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/91
- https://github.com/kiegroup/kogito-runtimes/pull/865
- https://github.com/kiegroup/optaplanner/pull/1010
- https://github.com/kiegroup/kogito-examples/pull/435
- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/114

### CODE REVIEW

File: Jenkinsfile.deploy
Comment (radtriste) [inline Jenkinsfile.deploy:267]:
```suggestion
        cmd += " --projects ${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy
Comment (radtriste) [inline Jenkinsfile.deploy:267]:
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy
Comment (radtriste) [inline Jenkinsfile.deploy:278]:
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy
Comment (radtriste) [inline Jenkinsfile.deploy:153]:
```suggestion
                    mavenCleanInstall('optaplanner', params.SKIP_TESTS)
```

**Comment by radtriste** [comment]
Almost there but testing is blocked by https://issues.redhat.com/browse/KOGITO-3795 ...

File: Jenkinsfile.promote
Comment (radtriste) [inline Jenkinsfile.promote:169]: Will be commented out once testing is finished

File: Jenkinsfile.promote
Comment (radtriste) [inline Jenkinsfile.promote:449]: Will be commented out once testing is finished

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:146]: The `mavenCleanInstallOptaPlannerParents` ran only specific modules (a parent pom.xml and BOM) the other projects depend upon to save time (a couple of seconds vs. ~5 minutes).
I don't expect these two modules to be renamed; especially the BOM is a part of our API.

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:226]: The `if-else` can be replaced by a method, e.g. `deployMavenArtifacts()` that would take care of a single repository.

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:303]: Thank you!

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:313]:
```suggestion
void runMavenDeploy(MavenCommand mvnCmd) {
```

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:316]:
```suggestion
    if(params.MAVEN_DEPLOY_REPOSITORY) {
```

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:342]:
```suggestion
String getLocalDeploymentFolder(String localDeployId) {
```

File: Jenkinsfile.deploy
Comment (rsynek) [inline Jenkinsfile.deploy:213]:
```suggestion
                    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY) {
```

File: Jenkinsfile.promote
Comment (rsynek) [inline Jenkinsfile.promote:468]:
```suggestion
void runMavenDeploy(MavenCommand mvnCmd) {
```

File: Jenkinsfile.deploy
Comment (radtriste) [reply Jenkinsfile.deploy]: something like:
```
deployMavenArtifacts(defaultOptaplannerMvnCmd, optaplannerRepository)
deployMavenArtifacts(defaultOptawebVehicleRoutingMvnCmd, vehicleRoutingRepository)
deployMavenArtifacts(defaultOptawebEmployeeRosteringMvnCmd, employeeRosteringRepository)
```
and then
```
void deployMavenArtifacts(mvnCmd, directory){
    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY){
         runMavenDeploy(mvnCmd)
    } else {
        runMavenDeployLocally(mvnCmd, directory)
        runMavenStage(mvnCmd, directory)
    }
}
```
?

File: Jenkinsfile.deploy
Comment (radtriste) [reply Jenkinsfile.deploy]: woops. gonna correct that

Comment (radtriste) [reply Jenkinsfile.deploy]: updated

File: .gitignore
Comment (rsynek) [inline .gitignore:17]: These are eclipse files? They should be already covered by `/.*` on the line 5.

File: .jenkins/tests/pom.xml
Comment (rsynek) [inline .jenkins/tests/pom.xml:2]: While I like the idea of having a PR check for jenkinsfiles, I have concerns about maintainability of this solution:
1/ it brings a lot of dependencies to the project (although it's a separate maven build)
2/ it duplicates the same (or very similar) setup in kogito-runtimes, kogito-examples and kogito-pipeline (in form of build.gradle) and maybe other repositories.

We should think about reducing the maintainability cost before merging the PRs.

First proposal: creating a parent pom.xml (in kogito-pipelines or jenkins-pipeline-shared-libraries ?) the other repositories would inherit from.

**Review by rsynek — CHANGES_REQUESTED** [review]
Please see my comment regarding the new pom.xml. Otherwise looks great.

File: .jenkins/tests/pom.xml
Comment (radtriste) [reply .jenkins/tests/pom.xml:2]: I see your concerns. I will remove the tests and gonna implement it as part of another issue (I need those PRs to be merged asap)
and I note your proposal for later implementation :)

Comment (radtriste) [reply .jenkins/tests/pom.xml]: https://issues.redhat.com/browse/BXMSPROD-1125

Comment (rsynek) [reply .jenkins/tests/pom.xml]: Perfect, thank you!

**Review by rsynek — APPROVED** [review]

**Comment by radtriste** [comment]
@rsynek can you review one last time and merge if ok ?

**Review by rsynek — APPROVED** [review]

**Pull request merged**
